File: bundles/org.eclipse.vjet.vsf.vjo.kernel/src/org/eclipse/vjet/vjo/util/List.java
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
package org.eclipse.vjet.vjo.util;

import java.util.ArrayList;
import org.eclipse.vjet.dsf.javatojs.anno.AJsProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsProxy;
import org.mozilla.mod.javascript.Scriptable;
import org.eclipse.vjet.dsf.javatojs.anno.AExclude;
import org.eclipse.vjet.dsf.dap.proxy.INativeJsFuncProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsFuncProxy;
import org.eclipse.vjet.dsf.dap.proxy.NativeJsTypeRef;
import org.eclipse.vjet.dsf.javatojs.anno.AJavaOnly;

//NativeJsProxy for vjo.util.List.js
@org.eclipse.vjet.dsf.resource.utils.CodeGen("NativeJsProxyGenerator")
@AJsProxy
public class List extends NativeJsProxy {

    /** for framework use only */
    @AExclude
    public List(Scriptable nativeObj){
        super(nativeObj);
    }

    /** internal use only */
    protected List(Object ...args){
        super(args);
    }

    public List(Object psJavaClass) {
        super(psJavaClass);
    }

    public List(ArrayList psJavaClass) {
        super(psJavaClass);
    }

    public Object get(int index) {
        return callWithName("get", Object.class, index);
    }

    public Object add(Object value) {
        return callWithName("add", Object.class, value);
    }

    public boolean remove(Object value) {
        return callWithName("remove", Boolean.class, value);
    }

    public int size() {
        return callWithName("size", Integer.class);
    }

    @AJavaOnly
    public static final NativeJsTypeRef<List> prototype = NativeJsTypeRef.get(List.class);

    public final INativeJsFuncProxy<List> get = NativeJsFuncProxy.create(this, "get");

    public final INativeJsFuncProxy<List> add = NativeJsFuncProxy.create(this, "add");

    public final INativeJsFuncProxy<List> remove = NativeJsFuncProxy.create(this, "remove");

    public final INativeJsFuncProxy<List> size = NativeJsFuncProxy.create(this, "size");
}

File: vsf/bundles/org.eclipse.vjet.vsf.vjo.kernel/bin/org/eclipse/vjet/vjo/util/List.js
/*******************************************************************************
 * Copyright (c) 2012 eBay Inc. and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     eBay Inc. - initial API and implementation
 *******************************************************************************/
/**
* Represents an list structrue used to handle response data.
*/
vjo.ctype('org.eclipse.vjet.vjo.util.List')
.protos({
	//> public void constructs(Object);
	//> public void constructs(java.util.ArrayList);
	constructs : function (psJavaClass) {
		this.javaClass = psJavaClass || 'java.util.ArrayList';
		this.list = [];
	},
	
	/**
	* Returns the element at the specified position in this list.
	*
	* @param {int} index 
	*        Index of element to return
	* @return {Object} 
	*        The element on the position
	*/
	//> public Object get(int);
	get : function (index) {
		if (this.size()>index) {
			return this.list[index]
		}
		//array out of bounds exception?
		return null;
	},
	
	/**
	* Appends the specified element to the end of this list (optional operation). 
	*
	* @param {Object} value 
	*        Element to be appended to this list
	* @return {Object} 
	*        The added element
	*/
	//> public Object add(Object);
	add : function (value) {
		return this.list[this.size()] = value;
	},
	
	/**
	* Removes the value if it in the list.
	* 
	* @param {Object} value 
	*        Element to be removed
	* @return {boolean} 
	*        True if the element in the list and be successfully removed
	*/
	//> public boolean remove(Object);
	remove : function (value) {//TODO: by index also?
		var len = this.size(), tmp = this.list, nlist = [], rv = false;
		for (var i=0; i<len; i++) {
			if (!rv && tmp[i]===value) {
				rv = true;
			} else {
				nlist[nlist.length] = tmp[i];
			}
		}
		this.list = nlist;
		return rv;
	},
	
	/**
	* Gets the length of current list.
	*
	* @return {int} 
	*         The length of current list 
	*/
	//> public int size();
	size : function () {
		return this.list.length;
	}
	//for all in?
})
.endType();